chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -596,29 +616,55 @@ impl<I: Interner> ToProgramClauses<I> for TraitDatum<I> {
             // conditions.
             let type_parameters: Vec<_> = trait_ref.type_parameters(interner).collect();
 
-            // Drop trait can't have downstream implementation because it can only
-            // be implemented with the same genericity as the struct definition,
-            // i.e. Drop implementation for `struct S<T: Eq> {}` is forced to be
-            // `impl Drop<T: Eq> for S<T> { ... }`. That means that orphan rules
-            // prevent Drop from being implemented in downstream crates.
-            if self.well_known != Some(WellKnownTrait::Drop) {
-                // Add all cases for potential downstream impls that could exist
-                for i in 0..type_parameters.len() {
+            if environment.has_compatible_clause(interner) {
+                // Note: even though we do check for a `Compatible` clause here,
+                // we also keep it as a condition for the clauses below, purely
```

Comment:
I think that makes sense -- conceptually, I think the solver should work the same if all clauses were being returned all the time, the clauses code should only decide to not generate non-applicable clauses.

Reply:
Yes, agreed